shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/Val.hs

```diff
@@ -171,3 +136,120 @@ scaledMinDeposit v (Coin mv)
     -- size of the UTxO entry (ie the space the scaled minUTxOValue deposit pays)
     utxoEntrySizeWithoutVal :: Integer
     utxoEntrySizeWithoutVal = inputSize + outputSizeWithoutVal
+
+addrHashLen :: Integer
+addrHashLen = 28
+
+uint :: Integer
+uint = 5
+
+assetIdLen :: Integer
+assetIdLen = 30 -- I have no idea what this is supposed to be
+
+-- ================================================================================================
+-- There are 2 basic instances of the Val class, and most other instances are built from them
+-- from some kind of automatic inheritance from these two types.
+
+instance Val Integer where
```

Comment:
No, there is no need for this. It unnecessarily exposes an implementation detail for no good reason. We use types for a purpose; allowing raw integers to accidentally be treated as `Val`s removes a layer of safety.

Reply:
You make some good points. I agree there are certainly some things to clean up. But I don't see it that way. Think of it this way. 

The Val class describes an algebraic structure. I agree that Val is probably not the best name.  Think of it as big box with a bunch of labelled items. There are two ways to label an item.
1) By path to the item.  
2) By associating a key to the item. 

There is an inductive way to define things in the Val class. We can generalize the discussion below to types other than Integer. It has three formation rules.

1) Base case:  instance Val Integer where
2) Path case:  instance (Val x, Val y) => instance Val (x,y) where
3) Key case: instance (Ord k,Val v) => instance Val (Map k v) where

This description has a bunch of algebraic properties, and this class exists independent of the uses we want to use it for. We can prove the properties using the inductive definition.  Thinking of it this way, the uses we want: Coin, and Value, get these properties for free, with a bit of help.

A) newtype deriving, allows to make a Val instance for any newtype whose embedded type is in the Val class.
B) Manual lifting of algebraic data types which have two fields both in the Val class to mimic the pair base path case.

So we see that Coin and Value are isomorphic to these instances of this inductive scheme.

Coin = Base
Value = Path Base (Key x (Key y Base))

We can describe the properties on the Val class in terms of its methods.
  zero :: t
  (<+>) :: t -> t -> t
  (<->) :: t -> t -> t
  (<×>) :: Integer -> t -> t
  invert :: t -> t
  isZero :: t -> Bool
  pointwise :: (Integer -> Integer -> Bool) -> t -> t -> Bool

The first set of properties show that minus (<->) and invert relate to scaling <×>

  x <-> y = x <+> (invert y)
  invert x = (-1) <×> x

We add second set of methods and properties that relate to the specific uses we have in mind. I suppose we could put these in a separate class. But having them all in 1 class adds simplicity and clarity.

a. That Coin is a blessed currency, and that every instance has an integer labeled Coin
b. That we need to compute the size of these things.
  
  coin :: t -> Coin
  inject :: Coin -> t
  modifyCoin :: (Coin -> Coin) -> t -> t
  size :: t -> Integer

The above methods apply to Both Coin and Value. There are additional operations and properties for Values.
    lookup :: PolicyID era -> AssetID -> Value era -> Integer
    insert :: (Integer -> Integer -> Integer) -> 
                 PolicyID era -> AssetID -> Integer -> Value era -> Value era